src/main/sensors/adcinternal.c

```diff
@@ -0,0 +1,112 @@
+/*
+ * This file is part of Cleanflight.
+ *
+ * Cleanflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "platform.h"
+
+#if defined(USE_ADC) && defined(USE_ADC_INTERNAL)
+
+#include "build/debug.h"
+
+#include "common/utils.h"
+
+#include "drivers/adc.h"
+
+typedef struct movingAverageStateUint16_s {
+    uint32_t sum;
+    uint16_t *values;
+    uint8_t size;
+    uint8_t pos;
+} movingAverageStateUint16_t;
+
+uint16_t updateMovingAverageUint16(movingAverageStateUint16_t *state, uint16_t newValue)
```

Comment:
Using existing BF implementation of IIR  filter will be simpler and provide comparable or better results.

Reply:
You mean FIR filters? IIRC, they all use `float`...

Reply:
IIR (https://en.wikipedia.org/wiki/Infinite_impulse_response). BF filters are implemented using float (integer/fixed point implementation is possible, but a bit tricky), with HW support that shouldn't be big problem.
Averaging filter is low-pass FIR filter (boxcar filter) with pretty poor frequency response. Main advantage is possible optimization, processing each sample only twice.
Using first order (PT1) or second order (biquad) filter will work quite well for ADC smoothing. One advantage is that cutoff frequency is specified, instead  of quite arbitrary number of averaged samples.

Reply:
BTW: here is comparison between 1st order IIR and moving average filter: https://dsp.stackexchange.com/a/14648